magma/backend/mlir/magma_common.py

```diff
@@ -67,18 +75,28 @@ def __iter__(self) -> Iterable['ValueWrapper']:
 def visit_value_or_value_wrapper_by_direction(
         value_or_value_wrapper: ValueOrValueWrapper,
         input_visitor: Callable[[Type], Any],
-        output_visitor: Callable[[Type], Any]):
+        output_visitor: Callable[[Type], Any],
+        **kwargs):
+
+    def descend(v):
+        if isinstance(v, (m_Tuple, Array)):
+            for item in v:
+                visit_value_or_value_wrapper_by_direction(
+                    item, input_visitor, output_visitor, **kwargs)
+            return
+        raise TypeError(value)
```

Comment:
```suggestion
        if not isinstance(v, (m_Tuple, Array)):
            raise TypeError(value)
        for item in v:
            visit_value_or_value_wrapper_by_direction(
                item, input_visitor, output_visitor, **kwargs)
```

Reply:
Could also be not a nested func

Reply:
good point -- doesn't strictly need to be, but it's easier to enclose the kwargs this way